// @aws-amplify/cli/lib/domain/amplify-usageData/UsageDataPayload.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SerializableError = exports.UsageDataPayload = void 0;
const os_1 = __importDefault(require("os"));
const VersionManager_1 = require("./VersionManager");
class UsageDataPayload {
    constructor(sessionUuid, installationUuid, version, input, error, state) {
        this.sessionUuid = sessionUuid;
        this.installationUuid = installationUuid;
        this.amplifyCliVersion = version;
        this.input = input;
        this.timestamp = new Date().toISOString();
        this.osPlatform = os_1.default.platform();
        this.osRelease = os_1.default.release();
        this.nodeVersion = process.versions.node;
        this.state = state;
        this.payloadVersion = VersionManager_1.getLatestPayloadVersion();
        if (error) {
            this.error = new SerializableError(error);
        }
    }
}
exports.UsageDataPayload = UsageDataPayload;
class SerializableError {
    constructor(error) {
        this.name = error.name;
    }
}
exports.SerializableError = SerializableError;
